system/smbus.c

```diff
@@ -62,6 +71,73 @@ static inline uint8_t bcd_to_ui8(uint8_t bcd)
     return bcd - 6 * (bcd >> 4);
 }
 
+// Returns integer part of parsed SPD5118 temperature in degree Celsius.
+static int16_t parse_spd5118_temperature(uint8_t high, uint8_t low) {
+    // Per JEDEC 300-5B standard (v1.5) SPD5 5118 Hub present on DDR5 modules
+    // store temperature values in range from -256.0C to 255.75C with 0.25C
+    // precision in two bytes (low, high) with the following bits layout:
+    //
+    // ----------+------+------+------+------+-----+------+------+------+
+    //      bit: |   7  |   6  |   5  |   4  |  3  |   2  |   1  |   0  |
+    // ----------+------+------+------+------+-----+------+------+------+
+    // high byte | RSVD | RSVD | RSVD | Sign | 128 |  64  |  32  |  16  |
+    //  low byte |   8  |   4  |   2  |   1  | 0.5 | 0.25 | RSVD | RSVD |
+    // ----------+------+------+------+------+-----+------+------+------+
+    //
+    // This is essentially 11-bit two-completent signed integer, e.g.
+    //   high=0b00000001 low=0b00001000 is 16.50 C
+    //   high=0b00000000 low=0b00001100 is  0.75 C
+    //   high=0b00000000 low=0b00000000 is  0.00 C
+    //   high=0b00011111 low=0b11111100 is -0.25 C
+    //   high=0b00011111 low=0b11111000 is -0.50 C
+
+    // Read 11-bit of two-complement value from high/low bytes into 16-bit
+    // variable.
+    int16_t temp4 = ((high & 0b11111) << 6) | ((low >> 2) & 0b111111);
+    // Extend sign from bit 11 to make it 16-bit signed integer.
+    temp4 = (temp4 << 4) >> 4;
```

Comment:
Offsets there should be 5, not 4. Well, I stated that the negative temperatures are not tested :)